### ISSUE TITLE
Intermittent errors while running tests after updating to 3.6

### ISSUE THREAD

**Issue by ChrisMaddock** [nunit/nunit-console#168]
@appel1 commented on [Mon Jan 30 2017](https://github.com/nunit/nunit/issues/2012)

Every now and then when we run our tests using the console runner after updating to 3.6 it fails with an error for some of the test assemblies. Tests for assemblies with this error are not discovered nor executed. It is not always the same assemblies that fails, sometimes it is one and sometimes there are more than one that fails. 

We execute the tests with: nunit3-console.exe [path to assemblies] -work=[absolute path to a directory] --result=TestResult.xml;format=nunit2

So far we've observed these errors with similar callstacks:
> Index was outside the bounds of the array.
>    at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
>    at NUnit.Engine.Services.TestAgency.LaunchAgentProcess(TestPackage package)
>    at NUnit.Engine.Services.TestAgency.CreateRemoteAgent(TestPackage package, Int32 waitTime)
>    at NUnit.Engine.Runners.ProcessRunner.CreateAgentAndRunner()
>    at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)

> Object reference not set to an instance of an object.
>    at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
>    at NUnit.Engine.Services.TestAgency.LaunchAgentProcess(TestPackage package)
>    at NUnit.Engine.Services.TestAgency.CreateRemoteAgent(TestPackage package, Int32 waitTime)
>    at NUnit.Engine.Runners.ProcessRunner.CreateAgentAndRunner()
>    at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)

---

@ChrisMaddock commented on [Mon Jan 30 2017](https://github.com/nunit/nunit/issues/2012#issuecomment-276062189)

I'll move this to the console repo, and comment over there. 😄

**Comment by ChrisMaddock** [comment]
Looks like the AgentDataBase nested class in TestAgency isn't threadsafe, and can be written to in parallel when multiple agents are created for test runs. I haven't looked into why this has only been a problem with 3.6.

**Comment by ChrisMaddock** [comment]
Update: I think was introduced with https://github.com/nunit/nunit-console/pull/141 - which removed the call to `Load()` in favour of just-in-time agent creation. The previous Load created agents processes sequentially, whereas now they're created inside RunTests, which runs in parallel, although the code to do so in TestAgency isn't thread-safe.  #141 was a fix to another bug itself - I haven't looked back any further to see what the previous state was!

What I understand from the commments on the PR - just-in-time agent creation and not loading packages at the start of the process are features we want. So instead we need to make sure everything under `ProcessRunner.CreateAgentAndRunner()` is thread-safe. @CharliePoole - you've been working here, are my assumptions correct? 😄

**Comment by CharliePoole** [comment]
@ChrisMaddock That sounds right. I'll look more closely to be sure. It's possible that we only need to make the data store that ties guids to processes threadsafe.

**Comment by CharliePoole** [comment]
@ChrisMaddock I'd appreciate if you'd review my conclusions in the code...

`ProcessRunner.CreateAgentAndRunner()` looks safe to me in and of itself. That pushes the problem down to `TestAgency.GetAgent(TestPackage, int)`.

Within `TestAgency`, the call sequence is `GetAgent => CreateRemoteAgent =>LaunchAgentProcess`

Since there is only one agency, those all need to be thread-safe. Both `CreateRemoteAgent` and `LaunchAgentProcess` use the agent database. In addition, when the agent calls back to register, the code in `TestAgency.Register` also uses it. I think making the agent database threadsafe is the key to everything here.

**Comment by ChrisMaddock** [comment]
That looks correct to me. The only other area, which I think we're already aware of, is RuntimeFramework - looks like CurrentFramework and AvailableFrameworks could both be subject to collision, if initialised at the same time. 

There's a hack in place to initialise AvailableFrameworks in RuntimeFrameworkService. CurrentFramework doesn't _look_ like it will ever cause a problem - but it could be good to explicitly initialise that in the same way! That said - I've just read the comment at the top of the RuntimeFramework class, that shows we're aware of all of this already - should we have an issue? 😄

**Comment by CharliePoole** [comment]
An issue wouldn't hurt. Most likely, we won't get as far as loading tests through `TestAgency` without having initialized CurrentFramework in the higher level code that doesn't involve parallelism. But it wouldn't hurt to be sure though.

**Comment by Trass3r** [comment]
I think we have the same issue. 3.5 is fine, 3.6 goes down though only on our CI build machines so far.
`nunit3-console.exe ..... --work=.../Release --out=TestOutput.log --result=TestResult.xml --labels=On --framework=net-4.5 --dispose-runners --agents=8`
```cs
The given key was not present in the dictionary.
   at System.ThrowHelper.ThrowKeyNotFoundException()
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at NUnit.Engine.Services.TestAgency.AgentDataBase.get_Item(Guid id)
   at NUnit.Engine.Services.TestAgency.CreateRemoteAgent(TestPackage package, Int32 waitTime)
   at NUnit.Engine.Services.TestAgency.GetAgent(TestPackage package, Int32 waitTime)
   at NUnit.Engine.Runners.ProcessRunner.CreateAgentAndRunner()
   at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)
```

**Comment by CharliePoole** [comment]
It's been sitting here as high priority for a while, without anybody taking it on. We can't do a hotfix without a fix. :smile:

**Comment by CharliePoole** [comment]
Nevertheless, I agree that it's critical.

**Comment by ChrisMaddock** [comment]
I was hoping I might get back to it this weekend. I feel like we've already done the difficult debugging work - and they actual fix will be about 2 lines long. 😆

**Comment by CharliePoole** [comment]
Ah! Hint taken. :smile:

**Comment by CharliePoole** [comment]
I've started working on this and will push another commit shortly. Not sure if I'll finish it however, so I'm leaving it in @ChrisMaddock 's name for now.

**Comment by CharliePoole** [comment]
Closed by #183

**Comment by ddobrev** [comment]
@CharliePoole we hit this one too with https://github.com/mono/CppSharp. Your latest release of 3.6.1 was uploaded on the 26th of February according to http://nunit.org/?p=download, i.e. before this issue was closed. Could you please tell if the fix actually entered 3.6.1 and if not, when it's going to be released?

**Comment by ChrisMaddock** [comment]
@ddobrev - this fix was required in the NUnit Console. The 3.6.1 release for that was 6th March - you're looking at the framework release date. 🙂
